feat: task scheduler

ryanolson opened the pull request:
I needed something similar for kvbm, so I made it general.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added robust management for critical background tasks, including monitoring, graceful shutdown, cancellation, and explicit failure propagation for improved reliability.
- Refactor
  - Reorganized runtime utilities into clearer modules without changing the public interface or behavior.
- Chores
  - Enabled additional runtime capabilities in async utilities to support improved task handling.
- Tests
  - Added comprehensive tests covering success, failure, cancellation, panics, timeouts, and lifecycle enforcement to ensure stability.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ryanolson committed:
rd 1

ryanolson committed:
improved metrics logic

ryanolson committed:
improve customizaton of children

ryanolson committed:
improve tests to control task concurrency directly instead of sleeps

ryanolson committed:
adding some convenience traits to beautify the api

ryanolson committed:
ready for review

PeaBrane commented on lib/runtime/src/utils/tasks/tracker.rs:
```diff
@@ -0,0 +1,3551 @@
+    }
+
+    /// Check if this tracker is closed
+    pub fn is_closed(&self) -> bool {
```
Does it make sense to also recursively is_closed for the children? But maybe not because of the guaranteed closing order enforced by `close()`

ryanolson replied:
Leaning toward the latter, if the present level is closed, then the children must be closed.

PeaBrane commented on lib/runtime/src/utils/tasks/tracker.rs:
```diff
@@ -0,0 +1,3551 @@
+    ///
+    /// Returns trackers in bottom-up order (children before parents) for safe shutdown.
+    /// This avoids recursive async calls and associated heap allocations.
+    fn collect_hierarchy(&self) -> Vec<Arc<TaskTracker>> {
```
Again nit. I don't think this would give a bottom-up order unless you are doing BFS. I believe this is still post-order

ryanolson replied:
updated comments.  not quite post-order.

nnshah1 commented on lib/runtime/src/utils/tasks/tracker.rs:
```diff
@@ -0,0 +1,3551 @@
+        self.local_metrics.active()
+    }
+
+    fn prometheus_metrics(&self) -> Option<Result<String>> {
```
would it make sense to leave this as a commented todo?

ryanolson replied:
this was a placeholder, I should remove it.  because we are creating the metrics objects thru the newly created dynamo metrics registry, the metrics object are already registered for us.

ryanolson committed:
Merge branch 'main' into ryan/task-scheduler

ryanolson commented on lib/runtime/src/utils/tasks/tracker.rs:
````diff
@@ -0,0 +1,3551 @@
+    /// println!("This tracker has {} active children", child_count);
+    /// # }
+    /// ```
+    pub fn child_count(&self) -> usize {
````
I was thinking of that, but we can always add that in if we need to.  For kvbm, each request slot will be a child since each request has some async onboarding and offloading tasks that it must complete before giving back ownership of its blocks.

ryanolson committed:
refactor(tasks): enforce cancellation semantics with split trait design

BREAKING CHANGE: Split TaskScheduler into ResourceGuard + acquisition phases

- Add ResourceGuard trait that enforces "always await completion"
- Update schedulers to separate resource acquisition from task execution
- Prevent cancellation misuse at compile time instead of documentation

ryanolson committed:
Merge branch 'ryan/task-scheduler' of github.com:ai-dynamo/dynamo into ryan/task-scheduler

ryanolson commented on lib/runtime/src/utils/tasks/tracker.rs:
```diff
@@ -0,0 +1,3551 @@
+}
+
+// Clone is automatically derived, but we need custom behavior for some fields
+impl Clone for TaskTracker {
```
I removed the Clone trait completely

ryanolson committed:
addressing some nits and confusing docstrings

ryanolson committed:
refactor

ryanolson committed:
fixed an issue where a cancellable task which was cancelled was reported as an error

ryanolson committed:
v2

ryanolson committed:
remove warnings

ryanolson committed:
checkpoint

ryanolson committed:
checkpoint 2

ryanolson committed:
restartable --> continuation

ryanolson committed:
getting closer

ryanolson committed:
OnErrorPolicy improvements

ryanolson committed:
allow for rescheduling of continuations

ryanolson committed:
last commit

ryanolson committed:
update doctests

ryanolson committed:
update a few more tests

rmccorm4 commented on lib/runtime/src/utils/tasks/tracker.rs:
@ryanolson to ease the review of 7k lines, can you add more details (or extract/distill some of the doc strings) to the PR description for a high level overview to reviewers?

rmccorm4 commented on lib/runtime/src/utils/tasks/tracker.rs:
```diff
@@ -0,0 +1,6285 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! # Task Tracker - Hierarchical Task Management System
```
For reviewers - you can run something like `cargo doc` on these changes to review all the examples in browser with syntax highlighting and such

ryanolson committed:
last docstring

keivenchang commented:
Some observations:

- TaskMetrics impl contains almost the _exact same methods_ as HierarchicalTaskMetrics trait. Both have 16 identical methods with the same signatures (only difference is pub vs fn keywords). The only difference is that TaskMetrics is missing the success() method that HierarchicalTaskMetrics requires. I suspect you want to enforce parity in both of them right? Means means using Trait on all of these Metrics structs.
- Question for you: what is the more intensive operation:
  - A) a program that frequently updates a bunch of RootTaskMetrics/DefaultRootTaskMetrics/ChildTaskMetrics, with periodic Prometheus fmt output? Or...
  - B) Very very frequent Prometheus fmt output, with infrequent incr()?
  - I suspect it would be A). If that's the case, it would be better to use AtomicU64 only. Then, during the infrequent Prometheus fmt output, copy cover to the Prometheus data-structure (gauge). In fact, you can increment the child (without the parent counter), and sum up the counters during these infrequent Prometheus fmt output.
- This RootTaskMetrics, DefaultRootTaskMetrics, and ChildTaskMetrics is very confusing to me.
  - RootTaskMetrics contains local_metrics as well as the Prometheus mirrors
  - DefaultRootTaskMetrics contains local_metrics
  - ChildTaskMetrics contains local_metrics and a parent, and the incr call will incr both
  - What do you think of simplying this such that there is just one node:
    - TaskMetricsNode, which contains TaskMetrics and the child TaskMetricsNode? It seems much more natural for nodes->[child_node1, child_node2, ...].
    - With a normal tree structure, you can still perform cancelled(), active(), success(), etc by traversing through the children and summing up the values. Traversing through the children periodically should be a lot cheaper than always counting twice in two separate data-structures.

I will go through your test cases and examples this weekend.

ryanolson committed:
simplified task metrics

ryanolson committed:
remove dead code

ryanolson committed:
Merge branch 'main' into ryan/task-scheduler

ryanolson committed:
Merge branch 'main' into ryan/task-scheduler

Signed-off-by: Ryan Olson <ryanolson@users.noreply.github.com>
